File: android/app/src/main/java/com/mochi/keyboard/data/FollowRepository.kt
package com.mochi.keyboard.data

import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.tasks.await

/**
 * `follows/{followerId}_{followeeId}` docs are client-writable, but `users/{uid}`'s
 * followerCount/followingCount fields are not (see firestore.rules) - those are maintained by a
 * Cloud Function fan-out trigger (WA3, not built yet). Same optimistic-UI caveat as LikeRepository.
 */
class FollowRepository(private val firestore: FirebaseFirestore) {

    private fun followDocId(followerId: String, followeeId: String) = "${followerId}_$followeeId"

    suspend fun isFollowing(followerId: String, followeeId: String): Boolean =
        firestore.collection("follows").document(followDocId(followerId, followeeId)).get().await().exists()

    suspend fun follow(followerId: String, followeeId: String) {
        val follow = hashMapOf("followerId" to followerId, "followeeId" to followeeId)
        firestore.collection("follows").document(followDocId(followerId, followeeId)).set(follow).await()
    }

    suspend fun unfollow(followerId: String, followeeId: String) {
        firestore.collection("follows").document(followDocId(followerId, followeeId)).delete().await()
    }

    /** Creator uids the given user follows - used to build the "Following" community feed tab. */
    suspend fun followedUids(followerId: String): List<String> =
        firestore.collection("follows")
            .whereEqualTo("followerId", followerId)
            .get()
            .await()
            .documents
            .mapNotNull { it.getString("followeeId") }
}

File: ios/MochiApp/Data/FollowRepository.swift
import FirebaseFirestore

/// `follows/{followerId}_{followeeId}` docs are client-writable, but `users/{uid}`'s
/// followerCount/followingCount fields are not — those are maintained by `onFollowWritten`
/// (functions/src/follows.ts). Same optimistic-UI caveat as LikeRepository, same contract
/// android/.../data/FollowRepository.kt uses.
final class FollowRepository {
    private let firestore: Firestore

    init(firestore: Firestore) {
        self.firestore = firestore
    }

    private func followDocId(followerId: String, followeeId: String) -> String { "\(followerId)_\(followeeId)" }

    func isFollowing(followerId: String, followeeId: String) async throws -> Bool {
        try await firestore.collection("follows").document(followDocId(followerId: followerId, followeeId: followeeId)).getDocument().exists
    }

    func follow(followerId: String, followeeId: String) async throws {
        try await firestore.collection("follows").document(followDocId(followerId: followerId, followeeId: followeeId))
            .setData(["followerId": followerId, "followeeId": followeeId])
    }

    func unfollow(followerId: String, followeeId: String) async throws {
        try await firestore.collection("follows").document(followDocId(followerId: followerId, followeeId: followeeId)).delete()
    }

    /// Creator uids the given user follows — used to build the "Following" community feed tab.
    func followedUids(followerId: String) async throws -> [String] {
        let snapshot = try await firestore.collection("follows").whereField("followerId", isEqualTo: followerId).getDocuments()
        return snapshot.documents.compactMap { $0.data()["followeeId"] as? String }
    }
}
